src/components/ListboxGrid/distribute-resources.ts

```diff
@@ -1,61 +1,232 @@
-    expanded = [...expanded, ...missingResource];
+  return columns;
+};
+
+export const balanceColumns = (size: ISize, columns: IColumn[]) => {
+  let collapsedItems = 0;
+  const expanded = columns.filter((column) => column.expand);
+  const collapsed = columns.filter((column) => column.expand === false);
+
+  // If there is room to expand one item in the last column of collapsed items we expand it.
+  // If not we distribute the items as evenly as possible among the columns with collapsed items.
+  if (
+    collapsed.length > 0
+    && !collapsed[collapsed.length - 1].showAll
+    && haveRoomToExpandOne(size, collapsed[collapsed.length - 1])
+  ) {
+    collapsed[collapsed.length - 1].expand = true;
+  } else {
+    collapsedItems = getColumnItemsCount(collapsed);
+    collapsed.forEach((column, index) => {
+      // If the dimensions can't be evenly distributed among the
+      // columns the extra dimensions should be given to the first columns
+      const extraItems = index < collapsedItems % collapsed.length ? 1 : 0;
+      const itemCount = Math.floor(collapsedItems / collapsed.length) + extraItems;
+
+      column.itemCount = itemCount;
+    });
+  }
+
+  return collapsed.concat(expanded);
+};
+
+export const mergeColumnsAndResources = (columns: IColumn[], resources: IListboxResource[]) => {
+  columns.forEach((column: IColumn) => {
+    column.items = resources.slice(0, column.itemCount);
+    resources = resources.slice(column.itemCount);
+  });
+
+  return columns;
+};
+
+export const expandOne = (sortedItems: IListboxResource[] | undefined, h: number) => {
+  if (!sortedItems) return false;
+  let i;
+  let item;
+  let expandedHeight;
+
+  for (i = 0; i < sortedItems.length; i++) {
+    item = sortedItems[i];
+    if (item.cardinal && !item.expand) {
+      expandedHeight = getHeightOfExpanded(item.cardinal);
+      if (expandedHeight < h) {
+        item.expand = true;
+        item.height = `${expandedHeight}px`;
+        return expandedHeight + ITEM_SPACING;
+      }
+    }
   }
 
-  return { expanded, folded, expandButton };
+  return false;
 };
+
+export const calculateExpandPriority = (columns: IColumn[], size: ISize) => {
+  columns.forEach((column: IColumn) => {
+    if (column.expand) {
+      let expandedCount = 0;
+      let leftOverHeight = size.height - getHeightOf((column.itemCount ?? 0) - 1) - ITEM_SPACING;
+      let totalExpandedHeight = 0;
+
+      // Make sure we have the cardinal so that they can be expanded in the best order
+      column?.items?.forEach((item) => {
+        item.cardinal = getDimensionCardinal(item);
+      });
+
+      const sortedItems = column?.items?.concat().sort((a, b) => a.cardinal - b.cardinal);
+
+      if ((sortedItems?.length ?? 0) > 1) {
+        while (leftOverHeight > EXPANDED_HEIGHT) {
+          const expHeight = expandOne(sortedItems, leftOverHeight);
+          if (expHeight) {
```

Comment:
What happens, and should happen, if `expHeight === 0`?

Reply:
Then the item is not expanded, it should go to `else`